### PR TITLE
fea(sql): add when condition for analytic function

### PR DESCRIPTION
**Pull request by superrxan**

### CODE REVIEW

File: internal/binder/function/funcs_analytic.go [hunk 1]
```diff
@@ -28,11 +28,11 @@ func registerAnalyticFunc() {
 	builtins["changed_col"] = builtinFunc{
 		fType: ast.FuncTypeScalar,
 		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
-			ignoreNull, ok := args[0].(bool)
-			if !ok {
-				return fmt.Errorf("first arg is not a bool but got %v", args[0]), false
+			if args[0].(bool) && args[1] == nil {
```
Comment (ngjaying) [inline internal/binder/function/funcs_analytic.go:31]: Maybe panic if args[0] is not a bool

File: internal/binder/function/funcs_analytic.go [hunk 2]
```diff
@@ -106,23 +107,25 @@ func registerAnalyticFunc() {
 	builtins["lag"] = builtinFunc{
 		fType: ast.FuncTypeScalar,
 		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
-			l := len(args) - 1
-			key := args[l].(string)
-			if l != 1 && l != 2 && l != 3 {
-				return fmt.Errorf("expect one two or three args but got %d", l), false
-			}
+			key := args[len(args)-1].(string)
```
Comment (ngjaying) [inline internal/binder/function/funcs_analytic.go:110]: The same problem as above

File: internal/binder/function/funcs_analytic.go [hunk 3]
```diff
@@ -106,23 +107,25 @@ func registerAnalyticFunc() {
...
 			v, err := ctx.GetState(key)
 			if err != nil {
 				return fmt.Errorf("error getting state for %s: %v", key, err), false
 			}
+			validData := args[len(args)-2].(bool)
```
Comment (ngjaying) [inline internal/binder/function/funcs_analytic.go:115]: same problem

File: internal/binder/function/funcs_analytic.go [hunk 4]
```diff
@@ -178,18 +178,17 @@ func registerAnalyticFunc() {
 	builtins["latest"] = builtinFunc{
 		fType: ast.FuncTypeScalar,
 		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
-			l := len(args) - 1
-			key := args[l].(string)
-			if l != 1 && l != 2 {
-				return fmt.Errorf("expect one or two args but got %d", l), false
-			}
+			paraLen := len(args) - 2
+			key := args[len(args)-1].(string)
```
Comment (ngjaying) [inline internal/binder/function/funcs_analytic.go:182]: same problem

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/binder/function/funcs_analytic.go (hunk 1 above)
Comment (superrxan) [reply internal/binder/function/funcs_analytic.go:31]: validate stage can check this error

File: internal/binder/function/funcs_analytic.go (hunk 2 above)
Comment (superrxan) [reply internal/binder/function/funcs_analytic.go:110]: validate stage can check this error

**Review by ngjaying — APPROVED** [review]

**Comment by ngjaying** [comment]
Closes #1498

**Pull request merged**